README.md
# Map Address Picker

[![Pub](https://img.shields.io/pub/v/map_address_picker.svg)](https://pub.dev/packages/map_address_picker)

A simple flutter plugin for picking users location using google maps. 


<p>
  <img src="https://raw.githubusercontent.com/athul-ain/map_address_picker/main/screenshots/Screenshot1.png" width=265/>
</p>

## Usage

To use this plugin, add `map_address_picker` as a dependency in your[ pubspec.yaml ](https://flutter.dev/docs/development/platform-integration/platform-channels) file.

## Getting Started

* Get an API key at <https://cloud.google.com/maps-platform/>.

* Enable Google Map SDK for each platform.
  * Go to [Google Developers Console](https://console.cloud.google.com/).
  * Choose the project that you want to enable Google Maps on.
  * Select the navigation menu and then select "Google Maps".
  * Select "APIs" under the Google Maps menu.
  * To enable Google Maps for Android, select "Maps SDK for Android" in the "Additional APIs" section, then select "ENABLE".
  * To enable Google Maps for iOS, select "Maps SDK for iOS" in the "Additional APIs" section, then select "ENABLE".
  * Make sure the APIs you enabled are under the "Enabled APIs" section.

For more details, see [Getting started with Google Maps Platform](https://developers.google.com/maps/gmp-get-started).

### Android

1. Set the `minSdkVersion` in `android/app/build.gradle`:

groovy
android {
    defaultConfig {
        minSdkVersion 20
    }
}


This means that app will only be available for users that run Android SDK 20 or higher.

2. Specify your API key in the application manifest `android/app/src/main/AndroidManifest.xml`:

```xml
<manifest ...
  <application ...
    <meta-data android:name="com.google.android.geo.API_KEY"
               android:value="YOUR KEY HERE"/>
```

### iOS

Specify your API key in the application delegate `ios/Runner/AppDelegate.m`:

```objectivec
#include "AppDelegate.h"
#include "GeneratedPluginRegistrant.h"
#import "GoogleMaps/GoogleMaps.h"

@implementation AppDelegate

- (BOOL)application:(UIApplication *)application
    didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
  [GMSServices provideAPIKey:@"YOUR KEY HERE"];
  [GeneratedPluginRegistrant registerWithRegistry:self];
  return [super application:application didFinishLaunchingWithOptions:launchOptions];
}
@end
```


Or in your swift code, specify your API key in the application delegate `ios/Runner/AppDelegate.swift`:

```swift
import UIKit
import Flutter
import GoogleMaps

@UIApplicationMain
@objc class AppDelegate: FlutterAppDelegate {
  override func application(
    _ application: UIApplication,
    didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
  ) -> Bool {
    GMSServices.provideAPIKey("YOUR KEY HERE")
    GeneratedPluginRegistrant.register(with: self)
    return super.application(application, didFinishLaunchingWithOptions: launchOptions)
  }
}
```


### Sample Usage

```dart
import 'package:map_address_picker/map_address_picker.dart';

LocationResult result = await showLocationPicker(context);

```


### Example 

```dart
import 'package:flutter/material.dart';
import 'package:map_address_picker/map_address_picker.dart';
import 'package:map_address_picker/models/location_result.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Map Address Picker Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      darkTheme: ThemeData.dark(),
      home: MyHomePage(title: 'Map Address Picker Example'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  LocationResult? locationResult;
  _openLocationPicker() async {
    var _result = await showLocationPicker(
      context,
      mapType: MapType.terrain,
      requiredGPS: true,
      automaticallyAnimateToCurrentLocation: true,
      // initialCenter: LatLng(28.612925, 77.229512),
      // desiredAccuracy: LocationAccuracy.best,
      // title: "Pick your location",
      // layersButtonEnabled: true,
      // initialZoom: 16,
      //floatingActionButtonLocation: FloatingActionButtonLocation.endFloat,
    );

    setState(() {
      locationResult = _result;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            ElevatedButton(
              onPressed: _openLocationPicker,
              child: Text("Pick Location"),
            ),
            Text(
              locationResult == null
                  ? "null"
                  : "${locationResult!.latLng!.latitude}\n${locationResult!.latLng!.longitude}",
              style: Theme.of(context).textTheme.headline4,
            ),
          ],
        ),
      ),
    );
  }
}

```


## Credits

The google map from [Flutter's](https://github.com/flutter) [google_maps_flutter](https://pub.dev/packages/google_maps_flutter) package

current location and permission from [BaseflowIT's](https://github.com/BaseflowIT) [flutter-geolocator](https://github.com/baseflowit/flutter-geolocator) package.

Inspired from [google_map_location_picker](https://github.com/humazed/google_map_location_picker) package.

CHANGELOG.md
## 0.3.5

* updated dependencies
* set compileSdkVersion to 33

## 0.3.0

* updated dependencies
* fixed incorrect location coordinates

## 0.0.8

* fixed incorrect location coordinates

## 0.0.5

* updated dependencies
* bug fixes

## 0.0.3

* fixed icon padding & changed button styles

## 0.0.1

* initial release.


lib/map_address_picker.dart
library map_address_picker;

export 'package:google_maps_flutter_platform_interface/src/types/ui.dart';
export 'package:google_maps_flutter_platform_interface/src/types/location.dart';
export 'package:google_maps_flutter_platform_interface/src/types/ui.dart';
export 'package:google_maps_flutter_platform_interface/src/types/location.dart';
export 'package:geolocator_platform_interface/src/enums/location_accuracy.dart';

import 'package:flutter/material.dart';
import 'package:geolocator/geolocator.dart';
import 'package:google_maps_flutter/google_maps_flutter.dart';
import 'package:map_address_picker/pages/location_picker.dart';
import 'models/location_result.dart';

/// Returns a [LatLng] object of the location that was picked.
///
/// [initialCenter] The geographical location that the camera is pointing
/// until the current user location is know if you want to change this
/// set [automaticallyAnimateToCurrentLocation] to false.
///
///
Future<LocationResult?> showLocationPicker(
  BuildContext context, {
  MapType mapType = MapType.normal,
  LatLng initialCenter = const LatLng(28.612925, 77.229512),
  bool requiredGPS = false,
  LocationAccuracy desiredAccuracy = LocationAccuracy.best,
  bool automaticallyAnimateToCurrentLocation = true,
  String? title,
  bool layersButtonEnabled = true,
  double initialZoom = 16,
  FloatingActionButtonLocation floatingActionButtonLocation =
      FloatingActionButtonLocation.endFloat,
}) async {
  final results = await Navigator.of(context).push(
    MaterialPageRoute<dynamic>(
      builder: (BuildContext context) {
        return LocationPickerPage(
          mapType: mapType,
          initialCenter: initialCenter,
          autoAnimateToCurrentLocation: automaticallyAnimateToCurrentLocation,
          desiredAccuracy: desiredAccuracy,
          requiredGPS: requiredGPS,
          title: title,
          layersButtonEnabled: layersButtonEnabled,
          initialZoom: initialZoom,
          floatingActionButtonLocation: floatingActionButtonLocation,
        );
      },
    ),
  );

  if (results != null && results.containsKey('location')) {
    return results['location'];
  } else {
    return null;
  }
}


lib/pages/location_picker.dart
import 'dart:async';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:geolocator/geolocator.dart';
import 'package:google_maps_flutter/google_maps_flutter.dart';
import 'package:map_address_picker/models/location_result.dart';

class LocationPickerPage extends StatefulWidget {
  final String? title;
  final MapType mapType;
  final LatLng initialCenter;
  final LocationAccuracy desiredAccuracy;
  final bool requiredGPS;
  final bool autoAnimateToCurrentLocation;
  final bool layersButtonEnabled;
  final double initialZoom;
  final FloatingActionButtonLocation floatingActionButtonLocation;
  const LocationPickerPage({
    Key? key,
    required this.mapType,
    required this.initialCenter,
    required this.desiredAccuracy,
    required this.requiredGPS,
    required this.autoAnimateToCurrentLocation,
    this.title,
    required this.layersButtonEnabled,
    required this.initialZoom,
    required this.floatingActionButtonLocation,
  }) : super(key: key);

  @override
  _LocationPickerPageState createState() => _LocationPickerPageState();
}

class _LocationPickerPageState extends State<LocationPickerPage> {
  Completer<GoogleMapController> mapController = Completer();
  Position? _currentPosition;
  LatLng? _lastMapPosition;
  MapType? selectedMapType;
  Set<Marker> _markers = {};

  // this also checks for location permission.
  Future<void> _initCurrentLocation() async {
    await _checkGeolocationPermission();
    Position? currentPosition;
    try {
      currentPosition = await Geolocator.getCurrentPosition(
        desiredAccuracy: widget.desiredAccuracy,
      );
    } catch (e) {
      currentPosition = null;
      print(e);
    }

    if (mounted && currentPosition != null)
      setState(() => _currentPosition = currentPosition);
    if (currentPosition != null || _currentPosition != null) {
      _lastMapPosition = currentPosition == null
          ? LatLng(_currentPosition!.latitude, _currentPosition!.longitude)
          : LatLng(currentPosition.latitude, currentPosition.longitude);
      setState(() {
        _markers.add(Marker(
          markerId: MarkerId("defaultMarker"),
          position: currentPosition == null
              ? LatLng(_currentPosition!.latitude, _currentPosition!.longitude)
              : LatLng(currentPosition.latitude, currentPosition.longitude),
        ));
      });
    }

    if (_currentPosition != null)
      await moveToCurrentLocation(
          LatLng(_currentPosition!.latitude, _currentPosition!.longitude));
  }

  Future moveToCurrentLocation(LatLng currentLocation) async {
    final controller = await mapController.future;
    controller.animateCamera(CameraUpdate.newCameraPosition(
      CameraPosition(target: currentLocation, zoom: widget.initialZoom),
    ));
  }

  @override
  void initState() {
    if (widget.autoAnimateToCurrentLocation && !widget.requiredGPS)
      _initCurrentLocation();
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    if (selectedMapType == null) selectedMapType = widget.mapType;

    if (widget.requiredGPS) {
      _checkGeolocationPermission();
      if (_currentPosition == null && widget.autoAnimateToCurrentLocation)
        _initCurrentLocation();
    }

    bool darkIcons = ((selectedMapType == MapType.hybrid) ||
        (selectedMapType == MapType.satellite));

    return Scaffold(
      appBar: AppBar(
        systemOverlayStyle: SystemUiOverlayStyle(
          statusBarColor: Colors.transparent,
          statusBarIconBrightness:
              darkIcons ? Brightness.light : Brightness.dark,
        ),
        backgroundColor: Colors.transparent,
        elevation: 0,
        iconTheme: Theme.of(context).iconTheme.copyWith(
              color: darkIcons ? Colors.white : null,
            ),
        title: Text(widget.title ?? ""),
        toolbarTextStyle: Theme.of(context)
            .textTheme
            .copyWith(
              headline6: TextStyle(
                color: darkIcons ? null : Colors.black,
              ),
            )
            .bodyText2,
        titleTextStyle: Theme.of(context)
            .textTheme
            .copyWith(
              headline6: TextStyle(
                color: darkIcons ? null : Colors.black,
              ),
            )
            .headline6,
      ),
      extendBodyBehindAppBar: true,
      body: Stack(
        children: [
          GoogleMap(
            mapType: selectedMapType ?? MapType.normal,
            initialCameraPosition: CameraPosition(
              target: widget.initialCenter,
              zoom: widget.initialZoom,
            ),
            onMapCreated: (GoogleMapController controller) {
              mapController.complete(controller);

              _markers.add(Marker(
                markerId: MarkerId("defaultMarker"),
                position: widget.initialCenter,
              ));

              _lastMapPosition = widget.initialCenter;
            },
            markers: _markers,
            onTap: (argument) {
              _lastMapPosition = argument;

              setState(() {
                _markers.add(Marker(
                  markerId: MarkerId("defaultMarker"),
                  position: argument,
                ));
              });
            },
            onLongPress: (argument) {
              _lastMapPosition = argument;

              setState(() {
                _markers.add(Marker(
                  markerId: MarkerId("defaultMarker"),
                  position: argument,
                ));
              });
            },
            onCameraIdle: () async {
              print("onCameraIdle#_lastMapPosition = $_lastMapPosition");
            },
            myLocationEnabled: true,
            myLocationButtonEnabled: false,
            padding: EdgeInsets.only(
              bottom: kBottomNavigationBarHeight + 30,
            ),
          ),
          Container(
            alignment: Alignment.topRight,
            margin: const EdgeInsets.only(top: kToolbarHeight + 30, right: 8),
            child: Column(
              children: [
                if (widget.layersButtonEnabled)
                  FloatingActionButton(
                    onPressed: () {
                      setState(() {
                        if (selectedMapType == MapType.normal) {
                          selectedMapType = MapType.hybrid;
                        } else {
                          selectedMapType = MapType.normal;
                        }
                      });
                    },
                    materialTapTargetSize: MaterialTapTargetSize.padded,
                    mini: true,
                    child: const Icon(Icons.layers),
                    heroTag: "layers",
                  ),
                if (widget.requiredGPS)
                  FloatingActionButton(
                    onPressed: _initCurrentLocation,
                    materialTapTargetSize: MaterialTapTargetSize.padded,
                    mini: true,
                    child: const Icon(Icons.gps_fixed),
                    heroTag: "layers2",
                  ),
              ],
            ),
          ),
        ],
      ),
      floatingActionButton: FloatingActionButton(
        child: Icon(Icons.arrow_forward),
        onPressed: () {
          Navigator.of(context).pop(
            {'location': LocationResult(latLng: _lastMapPosition)},
          );
        },
      ),
      floatingActionButtonLocation: widget.floatingActionButtonLocation,
    );
  }

  var dialogOpen;

  Future _checkGeolocationPermission() async {
    await Geolocator.requestPermission();
    final geolocationStatus = await Geolocator.checkPermission();
    print("geolocationStatus = $geolocationStatus");

    if (geolocationStatus == LocationPermission.denied && dialogOpen == null) {
      dialogOpen = _showDeniedDialog();
    } else if (geolocationStatus == LocationPermission.deniedForever &&
        dialogOpen == null) {
      dialogOpen = _showDeniedForeverDialog();
    } else if (geolocationStatus == LocationPermission.whileInUse ||
        geolocationStatus == LocationPermission.always) {
      print('GeolocationStatus.granted');

      if (dialogOpen != null) {
        Navigator.of(context, rootNavigator: true).pop();
        dialogOpen = null;
      }
    }
  }

  Future _showDeniedDialog() {
    return showDialog(
      context: context,
      barrierDismissible: false,
      builder: (context) {
        return WillPopScope(
          onWillPop: () async {
            Navigator.of(context, rootNavigator: true).pop();
            Navigator.of(context, rootNavigator: true).pop();
            return true;
          },
          child: AlertDialog(
            title: Text('Access to location denied'),
            content: Text('Allow access to the location services.'),
            actions: <Widget>[
              TextButton(
                child: Text('Ok'),
                onPressed: () {
                  Navigator.of(context, rootNavigator: true).pop();
                  _initCurrentLocation();
                  dialogOpen = null;
                },
              ),
            ],
          ),
        );
      },
    );
  }

  Future _showDeniedForeverDialog() {
    return showDialog(
      context: context,
      barrierDismissible: false,
      builder: (context) {
        return WillPopScope(
          onWillPop: () async {
            Navigator.of(context, rootNavigator: true).pop();
            Navigator.of(context, rootNavigator: true).pop();
            return true;
          },
          child: AlertDialog(
            title: Text(
              'Access to location permanently denied',
            ),
            content: Text(
              'Allow access to the location services for this App using the device settings.',
            ),
            actions: <Widget>[
              TextButton(
                child: Text('Ok'),
                onPressed: () {
                  Navigator.of(context, rootNavigator: true).pop();
                  Geolocator.openAppSettings();
                  dialogOpen = null;
                },
              ),
            ],
          ),
        );
      },
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:map_address_picker/map_address_picker.dart';
import 'package:map_address_picker/models/location_result.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Map Address Picker Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      darkTheme: ThemeData.dark(),
      home: MyHomePage(title: 'Map Address Picker Example'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  LocationResult? locationResult;
  _openLocationPicker() async {
    var _result = await showLocationPicker(
      context,
      mapType: MapType.terrain,
      requiredGPS: true,
      automaticallyAnimateToCurrentLocation: true,
      // initialCenter: LatLng(28.612925, 77.229512),
      // desiredAccuracy: LocationAccuracy.best,
      // title: "Pick your location",
      // layersButtonEnabled: true,
      // initialZoom: 16,
      //floatingActionButtonLocation: FloatingActionButtonLocation.endFloat,
    );

    if (mounted) setState(() => locationResult = _result);
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            ElevatedButton(
              onPressed: _openLocationPicker,
              child: Text("Pick Location"),
            ),
            Text(
              locationResult == null
                  ? "null"
                  : "${locationResult!.latLng!.latitude}\n${locationResult!.latLng!.longitude}",
              style: Theme.of(context).textTheme.headline4,
            ),
          ],
        ),
      ),
    );
  }
}
